File: Q0150_Evaluate_Reverse_Polish_Notation/Solution.java
package Q0150_Evaluate_Reverse_Polish_Notation;

import java.util.Stack;

/**
 * @ Description:
 * @ Date: Created in 19:48 2019-04-07
 * @ Author: Anthony_Duan
 */
public class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            if (token.equals("+") || token.equals("-") || token.equals("*") || token.equals("/")) {
                int a = stack.pop();
                int b = stack.pop();
                int re;
                switch (token) {
                    case "+":
                        re = a + b;
                        break;
                    case "-":
                        re = b - a;
                        break;
                    case "*":
                        re = a * b;
                        break;
                    case "/":
                        re = b / a;
                        break;
                    default:
                        throw new IllegalStateException("Unexpected value: " + token);
                }
                stack.push(re);
            } else {
                stack.push(Integer.valueOf(token));
            }
        }

        return stack.pop();
    }

    public static void main(String[] args) {
        String[] s = {"4", "-2", "/", "2", "-3", "-", "-"};
        System.out.println(new Solution().evalRPN(s));
    }

}

File: Q0150_Evaluate_Reverse_Polish_Notation/Solution.py
class Solution:
    def evalRPN(self, tokens) -> int:
        stack = []
        for i in tokens:
            if i in ["+", "-", "*", "/"]:
                a = stack.pop()
                b = stack.pop()
                if i is "/":
                    stack.append(str(int(eval(b + i + a))))
                else:
                    stack.append(str(eval(b + i + a)))

            else:
                stack.append(i)

        return int(stack.pop())


if __name__ == '__main__':
    print(Solution().evalRPN(["4", "-2", "/", "2", "-3", "-", "-"]))
